lib/OpenQA/WebAPI/Controller/Admin/Needle.pm

```diff
@@ -86,6 +95,12 @@ sub ajax {
                     last_seen  => $n->last_seen_time || 'never',
                     last_match => $n->last_matched_time || 'never',
                 };
+                my $key                   = $n->path;
+                my $needle_file_real_path = $needles_record_real_path{$key}->{needle_file_real_path};
+                if ($needle_file_real_path && $needles_record_real_path{$needle_file_real_path}) {
```

Comment:
I wouldn't repeat `$needles_record_real_path{$needle_file_real_path}` all the time.